File: workflow/scripts/screen_result.py
#!/usr/bin/env python3
import pandas as pd
from pathlib import Path

def read_data(vs2_path, checkV_path, diamond_path, output_path):
    # Read data files
    vs2 = pd.read_csv(vs2_path, sep='\t')
    checkV = pd.read_csv(checkV_path, sep='\t')
    diamond = pd.read_csv(diamond_path, sep='\t', skiprows=3, header=None)
    
    # Rename column names for diamond data
    diamond.columns = ["qseqid", "sseqid", "pident", "length", "mismatch", "evalue", 
                       "bitscore", "staxids", "sscinames", "sskingdoms", "skingdoms", 
                       "sphylums", "stitle"]
    
    # Filter vs2 data
    vs2_screened = vs2[(vs2['max_score'] > 0.5) & (vs2['length'] > 1000)]
    vs2_screened[['contig_id', 'gene']] = vs2_screened['seqname'].str.split(r'\|\|', expand=True)
    
    # Filter checkV data
    checkV_screened = checkV[(checkV['checkv_quality'].isin(["Low-quality", "Medium-quality", "High-quality"])) & 
                             (checkV['contig_length'] > 1000)]
    
    # Join vs2 and checkV data
    contigs_for_diamond = pd.merge(vs2_screened, checkV_screened, on="contig_id", how="inner")
    
    # Filter and sort diamond data
    diamond_combined = diamond[(diamond[2] > 30) & (diamond[6] > 50) & (diamond[3] > 30)]
    diamond_combined.columns = ["qseqid", "sseqid", "pident", "length", "mismatch", "evalue", 
                                "bitscore", "staxids", "sscinames", "sskingdoms", "skingdoms", 
                                "sphylums", "stitle"]
    
    # Perform left join
    screen_result = pd.merge(contigs_for_diamond, diamond_combined, on="contig_id", how="left")
    screen_result.drop_duplicates(subset='contig_id', keep="first", inplace=True)
    screen_result.sort_values(by=['checkv_quality', 'max_score_group', 'skingdoms', 'bitscore'], ascending=[True, True, False, False], inplace=True)
    
    # Save result to file
    screen_result.to_csv(output_path, index=False)

# Example usage:
vs2_path = "results/vs2/24_4_S19.vs2.final-viral-score.tsv"
checkV_path = "results/checkV/24_4_S19.checkv.quality_summary.tsv"
diamond_path = "results/diamond/24_4_S19.tsv.gz"
output_path = "results/screen_24_4_S19.csv"

read_data(vs2_path, checkV_path, diamond_path, output_path)

File: workflow/scripts/screen_result.R
#!/usr/bin/env Rscript
# trail #7
library(readr)
library(dplyr)
library(tidyverse)

read_data <- function(vs2_path, checkV_path, diamond_path, output_path) {
  vs2 <- read_tsv(vs2_path, col_names = TRUE)
  checkV <- read_tsv(checkV_path, col_names = TRUE)
  diamond <- read_tsv(diamond_path, skip = 3, col_names = FALSE, col_types = cols(), show_col_types = FALSE)
  
  # Rename column names for diamond data
  colnames(diamond) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "evalue", "bitscore", "staxids", "sscinames", "sskingdoms", "skingdoms", "sphylums", "stitle")
  
  # Filter vs2 data
  vs2_screened <- vs2 %>%
    filter(max_score > 0.5,
           length > 1000) %>%
    separate(seqname, into = c("contig_id", "gene"), sep = "\\|\\|")
  
  # Filter checkV data
  checkV_screened <- checkV %>%
    filter(checkv_quality %in% c("Low-quality", "Medium-quality", "High-quality"),
           contig_length > 1000)
  
  # Join vs2 and checkV data
  contigs_for_diamond <- full_join(vs2_screened, checkV_screened, by = "contig_id")
  
  # Filter and sort diamond data
  diamond_combined <- diamond %>%
    filter(pident > 30,
           bitscore > 50,
           length > 30) %>%
    arrange(desc(bitscore), evalue, desc(length), desc(pident)) %>%
    separate(qseqid, into = c("contig_id", "gene"), sep = "\\|\\|")
  
  # Perform left join
  screen_result <- left_join(contigs_for_diamond, diamond_combined, by = "contig_id") %>%
    distinct(contig_id, .keep_all = TRUE) %>% 
    arrange(checkv_quality, max_score_group, desc(skingdoms), desc(bitscore)) 
  
  # Save result to file
  write_csv(screen_result, output_path)
}

# example:
# read_data("results/vs2/24_4_S19.vs2.final-viral-score.tsv","results/checkV/24_4_S19.checkv.quality_summary.tsv","results/diamond/24_4_S19.tsv.gz", "results/24_4_S19_screened.csv")

vs2_path <- snakemake@input$vs2_path
checkV_path <- snakemake@input$checkV_path
diamond_path <- snakemake@input$diamond_path
output_path <- snakemake@output[1]


read_data(vs2_path, checkV_path, diamond_path, output_path)